### ISSUE TITLE
[0.3] Ingesting exchange bundle throws IndexError

### ISSUE THREAD

**Issue by reinka** [enigmampc/catalyst#42]
Dear Catalyst Maintainers,

Before I tell you about my issue, let me describe my environment:

# Environment

* Operating System: mac sierra
* Python Version: Python 2.7.13 :: Continuum Analytics, Inc.
* Python Bitness: 64
* How did you install Catalyst: conda
* Python packages: 

alembic                   0.9.6                     <pip>
backports.functools-lru-cache 1.4                       <pip>
bcolz                     0.12.1                    <pip>
Bottleneck                1.2.1                     <pip>
certifi                   2016.2.28                py27_0  
certifi                   2017.7.27.1               <pip>
chardet                   3.0.4                     <pip>
click                     6.7                       <pip>
contextlib2               0.5.5                     <pip>
cycler                    0.10.0                    <pip>
cyordereddict             1.0.0                     <pip>
Cython                    0.27.1                    <pip>
decorator                 4.1.2                     <pip>
empyrical                 0.2.1                     <pip>
enigma-catalyst           0.3                       <pip>
idna                      2.6                       <pip>
intervaltree              2.1.0                     <pip>
Logbook                   1.1.0                     <pip>
lru-dict                  1.1.6                     <pip>
Mako                      1.0.7                     <pip>
MarkupSafe                1.0                       <pip>
matplotlib                2.1.0                     <pip>
mkl                       2017.0.3                      0  
multipledispatch          0.4.9                     <pip>
networkx                  2.0                       <pip>
numexpr                   2.6.4                     <pip>
numpy                     1.13.1                   py27_0  
openssl                   1.0.2l                        0  
pandas                    0.19.2                    <pip>
pandas-datareader         0.5.0                     <pip>
patsy                     0.4.1                     <pip>
pip                       9.0.1                    py27_1  
pudb                      2017.1.4                  <pip>
Pygments                  2.2.0                     <pip>
pyparsing                 2.2.0                     <pip>
python                    2.7.13                        0  
python-dateutil           2.6.1                     <pip>
python-editor             1.0.3                     <pip>
pytz                      2017.2                    <pip>
readline                  6.2                           2  
requests                  2.18.4                    <pip>
requests-file             1.4.2                     <pip>
requests-ftp              0.3.1                     <pip>
scipy                     0.19.1              np113py27_0  
setuptools                36.4.0                   py27_1  
six                       1.11.0                    <pip>
sortedcontainers          1.5.7                     <pip>
SQLAlchemy                1.1.14                    <pip>
sqlite                    3.13.0                        0  
statsmodels               0.8.0                     <pip>
subprocess32              3.2.7                     <pip>
TA-Lib                    0.4.10                    <pip>
tables                    3.4.2                     <pip>
tk                        8.5.18                        0  
toolz                     0.8.2                     <pip>
urllib3                   1.22                      <pip>
urwid                     1.3.1                     <pip>
wheel                     0.29.0                   py27_0  
zlib                      1.2.11                        0  

Now that you know a little about me, let me tell you about the issue I am
having:

# Description of Issue

* What did you expect to happen?
To ingest poloniex data via `catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt`

* What happened instead?
$ catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt
Ingesting exchange bundle poloniex...
Exception ValueError: ValueError('chunkfile /Users/administrator/.catalyst/data/exchanges/poloniex/minute_bundle/00/35/003541.bcolz/close/data/__-21.blp not found',) in 'bcolz.carray_ext.carray.getitem_cache' ignored
    [=                                   ]  Fetching poloniex minute candles: :    3%
Traceback (most recent call last):
  File "/anaconda/envs/catalyst/bin/catalyst", line 11, in <module>
    sys.exit(main())
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/__main__.py", line 508, in ingest_exchange
    show_progress=show_progress
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 453, in ingest
    show_progress)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 431, in ingest_assets
    empty_rows_behavior='strip'
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 232, in ingest_ctable
    end_dt=end_dt
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bcolz.py", line 51, in load_raw_arrays
    return self._load_raw_arrays(fields, start_dt, end_dt, sids)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bcolz.py", line 54, in _load_raw_arrays
    start_idx = self._find_position_of_minute(start_dt)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/data/minute_bars.py", line 1209, in _find_position_of_minute
    False,
  File "catalyst/data/_minute_bar_internal.pyx", line 81, in catalyst.data._minute_bar_internal.find_position_of_minute
IndexError: Out of bounds on buffer access (axis 0)

Here is how you can reproduce this issue on your machine:
## Reproduction Steps

1. `catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt`
...

## What steps have you taken to resolve this already?

`$ ls -l /Users/administrator/.catalyst/data/exchanges/poloniex/minute_bundle/00/35/003541.bcolz/close/data/`

-rw-r--r--  1 apoehlmann  staff  4105 Oct 20 23:49 __0.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __1.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __10.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __11.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __12.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __13.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __14.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __15.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __16.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __17.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __18.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __19.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __2.blp
-rw-r--r--  1 apoehlmann  staff  3144 Oct 20 23:49 __20.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __3.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __4.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __5.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __6.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __7.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __8.blp
-rw-r--r--  1 apoehlmann  staff  4064 Oct 20 23:49 __9.blp

I then moved the data folder since it's already from the previous version but this didn't solve the problem, however the logs are now slightly different, the `Exception ValueError: ValueError('chunkfile [...])` part (beginning of the previous logs) is gone:

$ catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt
Ingesting exchange bundle poloniex...
    [=                                   ]  Fetching poloniex minute candles: :    3%
Traceback (most recent call last):
  File "/anaconda/envs/catalyst/bin/catalyst", line 11, in <module>
    sys.exit(main())
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/__main__.py", line 508, in ingest_exchange
    show_progress=show_progress
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 453, in ingest
    show_progress)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 431, in ingest_assets
    empty_rows_behavior='strip'
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bundle.py", line 232, in ingest_ctable
    end_dt=end_dt
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bcolz.py", line 51, in load_raw_arrays
    return self._load_raw_arrays(fields, start_dt, end_dt, sids)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange_bcolz.py", line 54, in _load_raw_arrays
    start_idx = self._find_position_of_minute(start_dt)
  File "/anaconda/envs/catalyst/lib/python2.7/site-packages/catalyst/data/minute_bars.py", line 1209, in _find_position_of_minute
    False,
  File "catalyst/data/_minute_bar_internal.pyx", line 81, in catalyst.data._minute_bar_internal.find_position_of_minute
IndexError: Out of bounds on buffer access (axis 0)

**Comment by fredfortier** [comment]
I was able to reproduce this issue. We are looking into it. In the meanwhile, using a range worked for me:
`catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt -s 2016-01-01`

I suspect that the problem lies with the 2015 bundles which is what we are investigating.

**Comment by fredfortier** [comment]
I can push the start date even earlier.

This works:
`catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt -s 2015-4-1`

But this fails:
`catalyst ingest-exchange -x poloniex -f minute -i xrp_usdt -s 2015-1-1`

We will get back to you shortly with a complete solution but I hope that this helps in the meanwhile.

**Comment by abnera** [comment]
Similar error on my end:

command: `catalyst ingest-exchange -x poloniex -f minute -s 2017-1-1 --show-progress`
Error:
```
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://s3.amazonaws.com/enigmaco/catalyst-bundles/exchange-poloniex/poloniex-minute-gno_btc-2017-01.tar.gz
```
If I specify a pair it works but it doesn't work without specifying a pair.

catalyst ingest-exchange -x poloniex -f minute -s 2017-1-1 --show-progress - i xrp_usdt

**Comment by lacabra** [comment]
This issue has been resolved with commit 3d88d6a2c707adfb4e3791773e1ace942f62f8aa in 0.3.3 release. Please upgrade to the latest version: `pip install enigma-catalyst --upgrade`. Thank you
